### PR TITLE
Functional IO package

### PR DESCRIPTION
**Pull request by patricksnape**
Now the IO package is much more functional and importers and exporters are just expected to be defined as functions/callables that take either a path (input) or a file handle (output) and do the appropriate job. This makes it much easier for users to leverage the menpo smart globbing etc but actually import whatever they want.

### TODO
  - [x] `mio.register_*_importer('.jpg', custom_jpg_importer)`

### CODE REVIEW

**Comment by patricksnape** [comment]
@jabooth Please review me :laughing: :rice_cracker: :ear_of_rice: :earth_africa: :banana:

File: menpo/io/input/base.py [hunk 1]
```diff
@@ -1,25 +1,31 @@
...
 from pathlib import Path
 import random
-from ..utils import _norm_path
-from menpo.base import menpo_src_dir_path, LazyList
+
+from menpo.base import menpo_src_dir_path, LazyList, partial_doc
+from menpo.compatibility import basestring
 from menpo.visualize import print_progress
 
+from ..utils import _norm_path, _possible_extensions_from_filepath, \
```
Comment (jabooth) [inline menpo/io/input/base.py]: can we do
```
from foo import (lots, of,
                 items)
```
not a fan of `\` + newline

File: menpo/io/input/base.py [hunk 2]
```diff
@@ -42,30 +48,157 @@ def data_path_to(asset_filename):
...
+        project.
+    """
+    return [p.name for p in data_dir_path().glob('*') if not p.is_dir()]
+
+
+def _register_importer(ext_map, extension, callable):
     r"""
-    Menpo's default image landmark resolver. Returns all landmarks found to have
+    Register a new importer for the given extension. This makes it much
+    simpler to add custom importers to the IO package.
```
Comment (jabooth) [inline menpo/io/input/base.py]: This docstring might read weird in a year (future menpo users won't know it was ever harder ;)

I'd drop the second sentence, and add an Example section if poss?

File: menpo/io/input/video.py [hunk 3]
```diff
@@ -5,96 +5,106 @@
...
-                    pass
-            # Only take the frames before the broken ones
-            ll = ll[:end+1]
-
-        return ll
+    import imageio
+
+    reader = imageio.get_reader(str(filepath), format='ffmpeg', mode='I')
+
+    def imageio_to_menpo(imio_reader, index):
```
Comment (jabooth) [inline menpo/io/input/video.py]: @patricksnape as discussed in person, either pull out of enclosing body (so it isn't a closure) or just used the closed over reader and don't take a parameter for clarity either way. Can be done after @JeanKossaifi adds his video reader though.

**Comment by jabooth** [comment]
@patricksnape really minor comments, but other than those looks really good. +1.

**Comment by patricksnape** [comment]
jenkins retest this please

**Comment by patricksnape** [comment]
Go home Appveyor you're drunk. Merging.

**Pull request merged**